added initial draft for upgrade to mesa (#1133)
o1-labs/docs2: docs/network-upgrades/mesa/index.mdx

Mesa bundles four Mina Improvement Proposals (MIPs) that change protocol behavior, plus two operational improvements to the upgrade flow itself. The canonical MIP specs live in the [MinaProtocol/MIPs](https://github.com/MinaProtocol/MIPs/tree/main/MIPS) repository; each section below links to the spec for the proposal it describes.

### Faster Blocks — [MIP6](https://github.com/MinaProtocol/MIPs/blob/main/MIPS/mip-0006-slot-reduction-90s.md)

Mesa halves Mina's slot time from **180 seconds to 90 seconds**, doubling block production frequency. To keep the long-term token emission rate flat, the per-block **coinbase reward is also halved** (from 720 MINA to 360 MINA). The total Mina supplied per epoch is unchanged.

Several second-order consequences flow from the slot-time change:

- **Epoch duration halves** from ~14.9 days to ~7.4 days. Scripts and operational procedures tied to epoch boundaries (delegation cooldown, automated payouts, monitoring dashboards) trigger about twice as often.
- **Vesting schedules on active vesting accounts are automatically migrated** during the hard fork so they continue unlocking on the same real-world cadence. No operator action is required.

> - **The default zkApp soft limit (24 commands per block) is unset**, allowing blocks to include more zkApp transactions when network demand is high.

Reviewer: This is not true. It actually went down temporarily; I kind of remember the other changes compensating for this, so total zkApp throughput didn't go down. I think it's 12 until we finish the transition frontier memory optimizations

Author: @cjjdespres, So removing ?

Author: Removed in `5df71880`

- **SNARK coordinators handling maximum-cost zkApp transactions should deploy at least 4 workers** (≈4 CPU cores each) to keep up with the new 90-second slot timing. A prerequisite SNARK-worker parallelization is delivered via soft fork before Mesa.

### Expanded zkApp State — [MIP7](https://github.com/MinaProtocol/MIPs/blob/main/MIPS/mip-0007-increase-state-size-limit.md)

Mesa raises the on-chain state available to a zkApp account from **8 field elements (indexes `0–7`) to 32 field elements (indexes `0–31`)**. zkApps can now store roughly four times more data directly on chain without off-chain workarounds.

Implications:

> - **zkApp developers must redeploy** to take advantage of the new fields; older verification keys will not validate transactions touching indexes `8–31`.

Reviewer: zkApp developers will need to redeploy regardless, I think?

Reviewer: I'm not sure this is the only reason for needing to redeploy, is what I mean

Author: @cjjdespres you want me to reword or remove that section. I don't mind either

Author: ok, removed as comment above is also pointing out issues with this section

Author: `5df71880`

- **Archive databases gain new columns** (`element8` … `element31`) on `zkapp_states` and `zkapp_states_nullable`. The Mesa archive upgrade script handles this automatically — see [Archive Node Schema Changes](/network-upgrades/mesa/appendix/archive-node-schema-changes) for the full schema diff.

> - **Transaction protocol version is bumped** because the on-chain serialization format changes.

Reviewer: There's also the zkApp permissions fallback logic change. Maybe the actions/events limit qualifies as a reason too? Anyway, could consider removing this from this section.

Reviewer: The transaction version bump, I mean

Author: removing implications at all

Author: `5df71880`

| **Exchanges** | Install the Mesa release. Keep deposits/withdrawals disabled until Mesa block production begins. |

---

### Phase 4: Post-Upgrade — after the fork

> **Goal:** Block production resumes on the Mesa network. Normal operations return.

At approximately **1 hour** after the Mesa release is published, at the predefined Mesa genesis timestamp, the first Mesa block is produced.

| Actor | What to do |
|---|---|
| **Block Producers** | Verify your node is on the Mesa chain (`mina-mesa client status`). If automode, check that the `activated` file exists in your config directory (see [Troubleshooting](/network-upgrades/mesa/upgrade-modes-details#how-do-i-know-which-binary-my-node-is-using)). Monitor block production. |
| **SNARK Workers** | Reconnect workers to the upgraded coordinator. Verify SNARK work is being produced. |
| **Archive Nodes** | Verify the archive database is in sync. Run [validation checks](/network-upgrades/mesa/upgrade-steps/post-upgrade#in-depth-validation). Fix any missing blocks using archive tooling. |
| **Exchanges** | **Re-enable MINA deposits and withdrawals** once block production is confirmed and your systems are verified. |

> | **zkApp Developers** | Deploy and test your upgraded zkApps on mainnet. Take advantage of the expanded 31-field on-chain state. |

Reviewer 2: 32-field

```bash
docker run --name rosetta --rm -d \
  -p 3088:3088 \
  --entrypoint '' \
  minaprotocol/mina-rosetta:4.0.0-bullseye-mainnet \
  /usr/local/bin/mina-rosetta \
    --archive-uri "postgres://<user>:<pass>@localhost:5432/archive_db" \
    --graphql-uri "http://localhost:3085/graphql" \
    --log-json --port 3088
```

> For detailed upgrade instructions, see [Upgrade](/network-upgrades/mesa/upgrade-steps/upgrade).

**After the fork (Post-Upgrade)** — Block production resumes. Eve verifies data integrity:

```bash
# Check the archive is in sync
mina client status

# Run the verification toolbox
mina-archive-hardfork-toolbox verify-upgrade \
  --postgres-uri postgres://<user>:<pass>@localhost:5432/archive_db \
  --protocol-version <version> \
  --migration-version <version>
```

She checks her explorer UI to confirm new Mesa blocks are appearing and the historical data is intact.

> For the full validation workflow, see [Post-Upgrade](/network-upgrades/mesa/upgrade-steps/post-upgrade) and the [Archive Hardfork Toolbox](/network-upgrades/mesa/archive-upgrade#verification-with-the-archive-hardfork-toolbox).

</details>

<details>
<summary><b>Example: zkApp Developer</b></summary>

> **Frank** maintains a zkApp deployed on mainnet. His contract uses on-chain state and he wants to take advantage of Mesa's expanded 31-field state.

Reviewer 2: 32-field

> ---

Reviewer 3: suggest to simplify this so most stuff go into subpage. This is too long.

title: Mesa Upgrade
sidebar_label: Mesa Upgrade
hide_title: true
description: Overview of the Mesa mainnet upgrade for Mina Protocol — expanded zkApp state, automode upgrades, and protocol improvements.
keywords:
  - Mesa
  - upgrade
  - hard fork
  - overview
  - mainnet
---

# Mesa Upgrade

The Mesa upgrade is a major network upgrade (hard fork) for the Mina Protocol mainnet. It is not backward compatible and requires all nodes to upgrade.

> Please refer to the Mesa node release notes [here](https://github.com/MinaProtocol/mina/releases).

Reviewer 3: Mesa release never happened, this must be a placeholder we need to fix later?

Author: Removing at all

:::info New to the Mesa upgrade?
This documentation uses terms like _automode_, _stop-slot_, _trustless upgrade_, _dispatcher_, and more. If you encounter an unfamiliar term, check the **[Glossary](/network-upgrades/mesa/glossary)** for definitions.
:::

> ## What Mesa Introduces

Reviewer 3: This didn't covered all MIPs we've introduced. For example account update limit bump is not being written here at all. Was it reverted?

Reviewer 2: Indeed, we are missing MIPs here. 
You can use the short overview describing each here: https://www.o1labs.org/blog/mesa-upgrade

### Expanded zkApp State

Mesa extends the on-chain state available to zkApps from 8 fields (indexes `0–7`) to 32 fields (indexes `0–31`) per account. This significantly increases the data capacity available to smart contracts, enabling more complex application logic without off-chain workarounds.

### Automode Upgrades

> For the first time in Mina's history, node operators can upgrade through a hard fork **without manual intervention**. The automode mechanism ships both the pre-fork and post-fork binaries in a single package, with a dispatcher that automatically transitions to the new chain when the fork activates. See [Upgrade Modes](/network-upgrades/mesa/upgrade-modes) for details.

Reviewer 3: It's not true. This only applies to daemon node. Suggest rewording so community will not be mad at us.

### Simplified Archive Upgrade

Unlike the Berkeley upgrade (which required up to 48 hours for archive database conversion), the Mesa archive upgrade is a fast schema upgrade that completes in under a minute. See [Archive Upgrade](/network-upgrades/mesa/archive-upgrade).

## Upgrade Flow — End-to-End Timeline

The Mesa upgrade follows four phases. The timeline below shows what happens at each stage, when it happens, and what **you** need to do.

### Overview

> <img src="/img/network-upgrades/mesa/simplified-schedule.png" alt="Mesa upgrade timeline overview" />

Reviewer 3: I would suggest removing such low-effort AI generated images. It make us looks unprofessional. Plus, I don't find useful information in this picture.

Reviewer: I think Rebecca is creating an updated version of the diagram in the Berkeley upgrade steps section: https://docs.minaprotocol.com/network-upgrades/berkeley/upgrade-steps

We can probably use that here instead.

Reviewer 2: yes, the new version is here, for example: https://docs.google.com/presentation/d/1y4RjpgPDfDruVKwPQdgJl38c2-wIPQr8rCydj_fwckk/edit?slide=id.g3df4c7c489f_0_103#slide=id.g3df4c7c489f_0_103

The upgrade moves through four phases — **Pre-Upgrade**, **State Finalization**, **Upgrade**, and **Post-Upgrade** — anchored by three key moments:

| Milestone | When | What happens |
|---|---|---|
| **stop-transaction-slot** | Hours before the fork | Network stops accepting new transactions |
| **stop-network-slot** | ~5 hours after stop-transaction-slot | Block production halts entirely |

> | **Mesa genesis timestamp** | ~1 hour after Mesa release is published | First Mesa block is produced |

Reviewer 3: Suggest @georgeee or @cjjdespres to confirm on the exact numbers here.

Reviewer: Yeah, stop transaction slot to stop network slot interval is exactly 5 hours, and stop network slot to mesa genesis is exactly 3 hours.

---

### Phase 1: Pre-Upgrade — weeks before the fork

> **Goal:** Every participant is prepared and running the stop-slot release before the fork begins.

| Actor | What to do |
|---|---|
| **All operators** | Verify [hardware requirements](/network-upgrades/mesa/requirements) (32 GB RAM, 8-core CPU with AVX/BMI2). Back up keys. |

> | **Block Producers** | Choose [upgrade mode](/network-upgrades/mesa/upgrade-modes): **automode** (recommended) or manual. Install stop-slot release [3.x.x](https://github.com/MinaProtocol/mina/releases). If automode, install both prefork and postfork packages. |

Reviewer 3: The instruction in this grid is specifically for debian users. I suggest pointing them to a separate page that people can read on different sections based on their setup.

Author: The grid no longer embeds Debian-specific install steps — the Quick Reference and Phase tables on this page now only link out (Requirements, Upgrade Modes), and the actual per-setup install commands live on the [Upgrade Modes](/network-upgrades/mesa/upgrade-modes) page (Debian / Docker tabs). So readers are pointed to setup-specific instructions rather than a Debian-only grid.

> | **Block Producers** | Choose [upgrade mode](/network-upgrades/mesa/upgrade-modes): **automode** (recommended) or manual. Install stop-slot release [3.x.x](https://github.com/MinaProtocol/mina/releases). If automode, install both prefork and postfork packages. |

Reviewer 2: Will the stop slot release be 3.5.0.?

> | **SNARK Workers** | Follow the same path as block producers (automode or manual). |

Reviewer 3: I don't know if there's a lot of people running REAL standalone snark workers, if there is, they need to update their snark workers separately. 

The "same path as block producer" claim works for snark coordinators(carrying a potential snark worker inside) only.

Reviewer: Yes, this was also discussed in other docs, that workers won't know exactly when the hard fork has happened and will need to be redeployed.

| **Archive Nodes** | Install stop-slot release. Choose upgrade method: _trustless_ (run upgrade script now) or _trustful_ (import o1Labs dump later). If trustless, run the [archive upgrade script](/network-upgrades/mesa/archive-upgrade) — it is backward compatible and can be applied early. |
| **Exchanges** | Install stop-slot release. Update integrations (mina-signer, Rosetta API). Test on devnet. Plan deposit/withdrawal freeze window. |

> | **zkApp Developers** | Update to the Mesa-compatible o1js version, recompile your contract, and verify it on the [preflight network](/network-upgrades/mesa/preflight-network). Plan to redeploy on Mesa after the fork — every zkApp must be redeployed because the protocol version bump changes the verification key. |

Reviewer: I think o1js will have upgrade docs - we should provide a link to them once they're out.

:::

---

### Phase 2: State Finalization — hours before the fork (~5 hours)

> **Goal:** The network reaches consensus on a final state. No new transactions are accepted.

At the predefined **stop-transaction-slot**, nodes stop accepting new user transactions. Block production continues for ~100 more slots with empty blocks (no coinbase, no fees) until the **stop-network-slot** to ensure all nodes converge on the same state.

| Actor | What to do |
|---|---|
| **Block Producers** | **Keep your node running.** Block density during finalization is critical. Do not stop your node. |
| **SNARK Workers** | **Keep running.** Continue producing SNARK work. |
| **Archive Nodes** | Keep the archive node running to capture all finalized blocks. If doing trustless upgrade, run the upgrade script now if you haven't already. |
| **Exchanges** | **Disable MINA deposits and withdrawals.** Any transactions submitted after the stop-transaction-slot will not exist on the Mesa chain. |
| **zkApp Developers** | No action required. Monitor announcements. |

:::danger For exchanges

> Transactions submitted after the stop-transaction-slot **will not carry over** to the Mesa chain. Freeze all MINA activity before this point.

Reviewer 3: Suggest adding a place holder page where we'll publish the specific numbers of fork time and slot-txn-end, slot-chain-end, etc. Exchanges should be able to be pointed to that page by clicking the "stop-transaction-slot" word here via a link, for example.

:::

---

### Phase 3: Upgrade — fork day (network is down)

> **Goal:** The network halts, state is exported, and the Mesa release is published.

At the **stop-network-slot**, block production stops entirely. o1Labs exports the network state, builds the Mesa release with the final ledger baked in, and publishes packages and Docker images.

| Actor | What to do |
|---|---|
| **Block Producers (automode)** | **Nothing.** Your node transitions to Mesa automatically. It will start producing blocks when the Mesa genesis timestamp arrives. |
| **Block Producers (manual)** | Stop your node. Wait for the Mesa release announcement. Install the new package. Restart with [updated flags](/network-upgrades/mesa/upgrade-steps/post-upgrade). |

> | **SNARK Workers** | Follow the same mode as your coordinator (automode = automatic, manual = reinstall and restart). |

Reviewer 3: Same as https://github.com/MinaProtocol/docs2/pull/1133/changes#r3198911314

| **Archive Nodes** | Install the Mesa archive node release. Point it at your upgraded database (trustless) or import the o1Labs SQL dump into a fresh database (trustful). |
| **Exchanges** | Install the Mesa release. Keep deposits/withdrawals disabled until Mesa block production begins. |

---

### Phase 4: Post-Upgrade — after the fork

> **Goal:** Block production resumes on the Mesa network. Normal operations return.

> At approximately **1 hour** after the Mesa release is published, at the predefined Mesa genesis timestamp, the first Mesa block is produced.

Reviewer 2: It's exactly 3 hours after release is published

| Actor | What to do |
|---|---|
| **Block Producers** | Verify your node is on the Mesa chain (`mina client status`). If automode, check that the `activated` file exists in your config directory (see [Troubleshooting](/network-upgrades/mesa/upgrade-modes-details#how-do-i-know-which-binary-my-node-is-using)). Monitor block production. |
| **SNARK Workers** | Reconnect workers to the upgraded coordinator. Verify SNARK work is being produced. |
| **Archive Nodes** | Verify the archive database is in sync. Run [validation checks](/network-upgrades/mesa/upgrade-steps/post-upgrade#in-depth-validation). Fix any missing blocks using archive tooling. |
| **Exchanges** | **Re-enable MINA deposits and withdrawals** once block production is confirmed and your systems are verified. |
| **zkApp Developers** | **Redeploy every zkApp** on Mesa — the protocol version bump invalidates Berkeley verification keys, so a redeploy is required before any zkApp can accept transactions. You can also take advantage of the expanded 32-field on-chain state at this point. |

---

> ### Timeline by Role

Reviewer 3: Not a fan of these pictures either.

Select your role to see a simplified view of what you need to do and when.

<details>

> <summary><b>Block Producers & SNARK Workers (automode)</b></summary>

Reviewer 3: again, we're mixing the concept of snark coordinator and snark workers. It's a bad thing that's being done in this doc before and somehow we're using AI to regenerate these hallucinations

</details>

<details>
<summary><b>Archive Node Operators</b></summary>

<img src="/img/network-upgrades/mesa/simplified-schedule-archive.png" alt="Archive node upgrade timeline" />

Choose your upgrade method early:
- **Trustless:** Run the archive upgrade script before the fork (it's backward-compatible). After the fork, point the Mesa archive node at your upgraded DB.
- **Trustful:** Wait for the o1Labs SQL dump after the fork and import it into a fresh database.

After the fork, verify data integrity and monitor the archive daemon.

</details>

<details>
<summary><b>Exchanges & Service Providers</b></summary>

<img src="/img/network-upgrades/mesa/simplified-schedule-echanges.png" alt="Exchange upgrade timeline" />

The critical action is timing your **deposit/withdrawal freeze**. Disable MINA activity before the stop-transaction-slot. Install the Mesa release on fork day. Re-enable deposits and withdrawals only after block production is confirmed and your integration is verified.

</details>

For step-by-step instructions, see [Upgrade Steps](/network-upgrades/mesa/upgrade-steps).

> ## Upgrade Modes

Reviewer 2: Remove this whole section from index. We already have the upgrade modes subpage and it's linked when we mention different modes. No need to repeat this info twice.

There are two ways to upgrade your daemon node:

- **[Automode](/network-upgrades/mesa/upgrade-modes) (Recommended):** The daemon handles the fork transition automatically. You install the stop-slot release and the node takes care of the rest — no manual intervention during the fork.
- **[Manual](/network-upgrades/mesa/upgrade-modes):** You manually stop your node after the network halts, install the Mesa release, and restart. This gives you full control.

For in-depth technical details on the dispatcher and dual-binary architecture, see [Upgrade Modes - Details](/network-upgrades/mesa/upgrade-modes-details).

> :::caution Automode requires a process restart and persistent filesystem

Reviewer 3: Why is this here instead of on upgrade-modes page?

Reviewer 2: +1 
This should be in upgrade-modes page

This means two things are required:
- **Persistent filesystem**: The config directory (e.g., `~/.mina-config` or the Docker volume at `/root/.mina-config`) **must survive across restarts**. If it is wiped, the `activated` file and generated Mesa config are lost, and the node cannot transition to Mesa.
- **Automatic restart policy**: Your process manager must be configured to restart the daemon after a clean exit:
  - **systemd**: Use `Restart=always` in your service file (this is the default in the Mina systemd unit)
  - **Docker**: Use `--restart=always` or `--restart=unless-stopped` when creating the container
  - **Kubernetes / Helm**: Ensure your pod has `restartPolicy: Always` (the k8s default). If you use a custom `livenessProbe` or a Helm chart with its own restart logic, verify it does not treat exit code 0 as a failure that triggers a full volume wipe or pod replacement.
:::

> :::info Dispatcher limitation (current implementation)

Reviewer 3: this should also go to subpage.

The automode dispatcher only supports the `daemon` subcommand and `mina client status` command. Other subcommands (e.g. `mina accounts list`, `mina ledger export`) will fail with an error. This is because, for non-daemon commands, the dispatcher does not have access to the config directory location and therefore cannot determine whether the node is running Berkeley or Mesa. It is the user's responsibility to invoke the correct version-specific binary directly: `mina-berkeley` or `mina-mesa` (or the full paths `/usr/lib/mina/berkeley/mina` and `/usr/lib/mina/mesa/mina`). This limitation may be removed in a future release.
:::

---

> ### Upgrade Examples

Reviewer 2: Remove this section from index.

Consolidate with the examples in the Upgrade Steps page and leave there.
Or, alternatively, consolidate these with the ones in Upgrade Steps and move to an "Examples" subpage of Upgrade Steps

The examples below walk through the entire upgrade for different roles and setups. Each follows the same four phases. Click to expand.

<details>
<summary><b>Example: Block Producer Upgrading with Automode (Debian)</b></summary>

Imagine you are **Alice**, a block producer running on a Debian server. Here is what your upgrade looks like end to end.

**Weeks before the fork** — Alice checks [hardware requirements](/network-upgrades/mesa/requirements) and installs the automode packages:

```bash
sudo apt-get update

> sudo apt-get install mina-mainnet-prefork-mesa=3.x.x mina-mainnet-postfork-mesa=4.x.x

Reviewer 2: Are we leaving the 3.x.x for now or do we know if we can already write it as 3.5.0 and 4.0.0?

```

She starts her node normally. The automode dispatcher runs the pre-fork binary until the fork activates.

> For full pre-upgrade checklist, see [Pre-Upgrade](/network-upgrades/mesa/upgrade-steps/pre-upgrade).

**Hours before the fork (State Finalization)** — The network reaches the _stop-transaction-slot_. Alice's node keeps producing blocks — she does **nothing**. Empty blocks are produced for ~100 slots (~5 hours) until all nodes agree on the final state.

> For details on this phase, see [State Finalization](/network-upgrades/mesa/upgrade-steps/state-finalization).

**Fork day (Upgrade)** — The network halts at the _stop-network-slot_. Alice's daemon generates the Mesa configuration, writes the `activated` marker file, and **shuts down cleanly** (exit code 0). Because Alice uses systemd with `Restart=always`, the daemon restarts automatically. On restart, the dispatcher detects the `activated` file and launches the Mesa binary. Alice does **nothing** — this all happens automatically.

> > For what happens if you chose manual mode instead, see [Upgrade](/network-upgrades/mesa/upgrade-steps/upgrade). For details on the restart mechanism, see [Upgrade Modes - Details](/network-upgrades/mesa/upgrade-modes-details#restart-and-filesystem-requirements).

Reviewer 2: Remove this, there is already another example that covers manual mode.

> **After the fork (Post-Upgrade)** — About 1 hour after the Mesa release is published, the first Mesa block is produced. Alice verifies:

Reviewer 2: ```suggestion
**After the fork (Post-Upgrade)** — About 3 hours after the Mesa release is published, the first Mesa block is produced. Alice verifies:
```

```bash
# Check if the activated file exists (path depends on your config directory and network ID)
ls ~/.mina-config/auto-fork-mesa-mainnet/activated

# Confirm Mesa chain ID
mina client status
```

> :::note Dispatcher and non-daemon commands

Reviewer 2: Not sure this should be here. This is an example, and Carlos does not have disclaimer, btw. I'd just move it to the Post-Upgrade step

**After the fork (Post-Upgrade)** — Block production resumes. Eve verifies data integrity:

```bash
# Check the archive is in sync
mina client status

# Run the verification toolbox
mina-archive-hardfork-toolbox verify-upgrade \
  --postgres-uri postgres://<user>:<pass>@localhost:5432/archive_db \
  --protocol-version <version> \
  --migration-version <version>
```

She checks her explorer UI to confirm new Mesa blocks are appearing and the historical data is intact.

> For the full validation workflow, see [Post-Upgrade](/network-upgrades/mesa/upgrade-steps/post-upgrade) and the [Archive Hardfork Toolbox](/network-upgrades/mesa/archive-upgrade#verification-with-the-archive-hardfork-toolbox).

</details>

<details>
<summary><b>Example: zkApp Developer</b></summary>

**Frank** maintains a zkApp deployed on mainnet. His contract uses on-chain state and he wants to take advantage of Mesa's expanded 32-field state.

**Weeks before the fork** — Frank updates his o1js dependency to the Mesa-compatible version and tests his zkApp on the [preflight network](/network-upgrades/mesa/preflight-network):

```bash

> npm install o1js@latest

Reviewer 2: latest is not accurate, as 3.0.0 is in pre-release. Latests would give you 2.15 if you are in pre-mesa.
If we expect them to use pre-flight it should now be o1js@3.0.0-mesa.698ca

```bash
mina client status   # verify Mesa chain ID
# test a small internal transfer before opening to customers
```

> For post-upgrade flags and configurations, see [Post-Upgrade](/network-upgrades/mesa/upgrade-steps/post-upgrade).

</details>

## Quick Reference by Operator Type

| Operator Type | Key Pages |
|---|---|
| **Block Producers** | [Requirements](/network-upgrades/mesa/requirements), [Upgrade Modes](/network-upgrades/mesa/upgrade-modes), [Upgrade Steps](/network-upgrades/mesa/upgrade-steps), [Post-Upgrade Flags](/network-upgrades/mesa/upgrade-steps/post-upgrade) |
| **SNARK Workers / Coordinators** | [Requirements](/network-upgrades/mesa/requirements), [Upgrade Steps](/network-upgrades/mesa/upgrade-steps), [Post-Upgrade Flags](/network-upgrades/mesa/upgrade-steps/post-upgrade) |
| **Archive Node Operators** | [Requirements](/network-upgrades/mesa/requirements), [Archive Upgrade](/network-upgrades/mesa/archive-upgrade), [Upgrade Steps](/network-upgrades/mesa/upgrade-steps), [Post-Upgrade](/network-upgrades/mesa/upgrade-steps/post-upgrade) |

> | **Rosetta API Operators** | [Archive Upgrade](/network-upgrades/mesa/archive-upgrade), [Upgrade Steps](/network-upgrades/mesa/upgrade-steps), [Post-Upgrade Flags](/network-upgrades/mesa/upgrade-steps/post-upgrade), [Post-Upgrade](/network-upgrades/mesa/upgrade-steps/post-upgrade) |

Reviewer 2: ```suggestion
| **Rosetta API Operators** | [Requirements](/network-upgrades/mesa/requirements), [Archive Upgrade](/network-upgrades/mesa/archive-upgrade), [Upgrade Steps](/network-upgrades/mesa/upgrade-steps), [Post-Upgrade](/network-upgrades/mesa/upgrade-steps/post-upgrade) |
```

> | **Exchanges** | [Requirements](/network-upgrades/mesa/requirements), [Upgrade Steps](/network-upgrades/mesa/upgrade-steps), [Appendix](/network-upgrades/mesa/appendix) |

Reviewer 2: ```suggestion
| **Exchanges** | [Requirements](/network-upgrades/mesa/requirements), [Upgrade Steps](/network-upgrades/mesa/upgrade-steps), [Post-Upgrade] (/network-upgrades/mesa/upgrade-steps/post-upgrade), [Appendix](/network-upgrades/mesa/appendix) |
```

> ## Network Details

Reviewer 3: Is this pre or postfork? Is this something we need to update when RC is released?
